tracker/progress.go

```diff
@@ -207,11 +207,13 @@ func (pr *Progress) MaybeDecrTo(rejected, matchHint uint64) bool {
 
 	// The rejection must be stale if "rejected" does not match next - 1. This
 	// is because non-replicating followers are probed one entry at a time.
+	// The check is a best effort assuming message reordering is rare.
 	if pr.Next-1 != rejected {
 		return false
 	}
 
-	pr.Next = max(min(rejected, matchHint+1), 1)
+	// Next index shall always be larger than match index.
+	pr.Next = max(min(rejected, matchHint+1), pr.Match+1)
```

Comment:
Did the test fail before this change? I.e. you verified that the bug was real?

Reply:
Yes, the test will fail at the last step of verification without this change.

Reply:
The test will fail without this change. The issue was actually found in trace validation when I enforce the model to always send MsgApp with larger match index. Then I reproduce this with this test case.